BraveShared/Analytics/UserReferralProgram.swift

```diff
@@ -189,27 +200,64 @@ public class UserReferralProgram {
+        
+        do {
+            return try NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData) as? [CustomHeaderData]
+        } catch {
+            log.error("Failed to unwrap custom headers with error: \(error.localizedDescription)")
+        }
+        return nil
     }
     
     /// Checks if a custom header should be added to the request and returns its value and field.
     public class func shouldAddCustomHeader(for request: URLRequest) -> (value: String, field: String)? {
-        guard let customHeadersAsData = Preferences.URP.customHeaderData.value,
-            let customHeaders = (try? NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData)) as? [CustomHeaderData],
+        guard let customHeaders = UserReferralProgram()?.customHeaders,
             let hostUrl = request.url?.host else { return nil }
         
         for customHeader in customHeaders {
             // There could be an egde case when we would have two domains withing different domain groups, that would
             // cause to return only the first domain-header it approaches.
-            for domain in customHeader.domainList {
-                if hostUrl.contains(domain) {
-                    if let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
-                        return (customHeader.headerValue, customHeader.headerField)
-                    }
+            for domain in customHeader.domainList where hostUrl.contains(domain) {
+                // If `domain` is "cookie only", we exclude it from HTTP Headers, and just use cookie approach
+                let cookieOnly = urpCookieOnlyDomains.contains(domain)
+                if !cookieOnly, let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
+                    return (customHeader.headerValue, customHeader.headerField)
                 }
             }
         }
         
         return nil
     }
+    
+    public class func insertCookies(intoStore store: WKHTTPCookieStore) {
+        guard let customHeadersDeferred = UserReferralProgram()?.fetchCustomHeaders() else { return }
+        
+        customHeadersDeferred.upon { customHeaders in
+            var cookies: [HTTPCookie] = []
+            customHeaders.forEach { header in
+                let domains = header.domainList.compactMap { URL(string: $0)?.absoluteString }
+                cookies += domains.compactMap {
+                    HTTPCookie(properties: [
+                        // Must include `.` prefix to be included in subdomains
+                        .domain: ".\($0)",
+                        .path: "/",
+                        .name: header.headerField,
+                        .value: header.headerValue,
+                        .secure: "TRUE",
+                        .expires: NSDate(timeIntervalSinceNow: 7.days)
```

Comment:
Do we expect the cookies to be accessed by scripts on coinbase? If not, we should probably add HTTPOnly flag as well: https://www.owasp.org/index.php/HttpOnly

Reply:
Unfortunately that flag is `readonly`, and not settable during the creation of the cookie (like all of these other flags).

Reply:
Please move discussion to: https://github.com/brave/brave-ios/pull/823/files